api/core/evaluate.js

```diff
@@ -28,82 +27,74 @@ export const execute = async (
-    let somethingFailed;
-    let parallelUpdates = 0;
-    const schedule = () => {
-      console.log(`Updates remaining ${[...pending].join(', ')}`);
-      for (const id of [...pending]) {
-        if (parallelUpdates >= parallelUpdateLimit) {
-          break;
+    let replaysDone = false;
+    console.log(`QQ/Evaluate`);
+    const scheduled = new Set();
+    for (;;) {
+      console.log(`QQ/Compile`);
+      const updates = {};
+      const replays = [];
+      const exports = [];
+      await toEcmascript(script, {
+        path,
+        topLevel,
+        updates,
+        replays,
+        exports,
+      });
+      // Replay anything we can.
+      if (!replaysDone) {
+        console.log(`QQ/Replay`);
+        replaysDone = true;
+        for (const entry of replays) {
+          await replay(entry, { path });
+        }
+      }
+      // Update what we can.
+      console.log(`QQ/Update`);
+      const blockedUpdates = [];
+      const updatePromises = [];
+      // Determine the updates we can process.
+      for (const id of Object.keys(updates)) {
+        if (scheduled.has(id)) {
+          continue;
         }
         const entry = updates[id];
         const outstandingDependencies = entry.dependencies.filter(
-          (dependency) =>
-            updates[dependency] &&
-            !processed.has(dependency) &&
-            dependency !== id
+          (dependency) => updates[dependency] && dependency !== id
         );
-        if (outstandingDependencies.length === 0) {
-          parallelUpdates++;
-          console.log(`Scheduling: ${id}`);
-          pending.delete(id);
+        if (
+          updatePromises.length === 0 &&
+          outstandingDependencies.length === 0
+        ) {
+          // For now, only do one thing at a time, and block the remaining updates.
           const task = async () => {
             try {
-              await evaluate(updates[id].program);
+              await evaluate(updates[id].program, { path });
               console.log(`Completed ${id}`);
-              delete updates[id];
-              unprocessed.delete(id);
-              processed.add(id);
-              parallelUpdates--;
             } catch (error) {
-              somethingFailed(error); // FIX: Deadlock?
-            } finally {
-              somethingHappened();
+              throw error;
             }
           };
-          addPending(task());
+          updatePromises.push(task());
+          scheduled.add(id);
+        } else {
+          blockedUpdates.push(id);
         }
       }
-    };
-    while (unprocessed.size > 0) {
-      const somethingHappens = new Promise((resolve, reject) => {
-        somethingHappened = resolve;
-        somethingFailed = reject;
-      });
-      schedule();
-      if (unprocessed.size > 0) {
-        // Wait for something to happen.
-        await somethingHappens;
+      // FIX: We could instead use Promise.race() and then see what new updates could be queued.
```

Comment:
Promise.race() is a neat feature that I didn't know about. That's a good one to know.